bpf/lib/fib.h

```diff
@@ -8,134 +8,181 @@
 #include <bpf/api.h>
 
 #include "common.h"
+#include "neigh.h"
 #include "l3.h"
 
-#ifdef ENABLE_IPV6
 static __always_inline int
-redirect_direct_v6(struct __ctx_buff *ctx __maybe_unused,
-		   int l3_off __maybe_unused,
-		   struct ipv6hdr *ip6 __maybe_unused, int *oif)
+maybe_add_l2_hdr(struct __ctx_buff *ctx __maybe_unused,
+		 __u32 ifindex __maybe_unused,
+		 bool *l2_hdr_required __maybe_unused)
+{
+	if (IS_L3_DEV(ifindex)) {
+		/* The packet is going to be redirected to L3 dev, so
+		 * skip L2 addr settings.
+		 */
+		*l2_hdr_required = false;
+	} else if (ETH_HLEN == 0) {
+		/* The packet is going to be redirected from L3 to L2
+		 * device, so we need to create L2 header first.
+		 */
+		__u16 proto = ctx_get_protocol(ctx);
+
+		if (ctx_change_head(ctx, __ETH_HLEN, 0))
+			return DROP_INVALID;
+		if (eth_store_proto(ctx, proto, 0) < 0)
+			return DROP_WRITE_ERROR;
+	}
+	return 0;
+}
+
```

Comment:
Also something that was needed for https://github.com/cilium/cilium/pull/23935/ :+1:.